File: 46.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


import bisect

class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def min_changes_to_bst(root):
    inorder_vals = []

    def inorder(node):
        if not node:
            return
        inorder(node.left)
        inorder_vals.append(node.val)
        inorder(node.right)

    inorder(root)
    n = len(inorder_vals)

    lis = []

    for val in inorder_vals:
        idx = bisect.bisect_left(lis, val)
        if idx == len(lis):
            lis.append(val)
        else:
            lis[idx] = val

    return n - len(lis)


# Example usage:
# This tree violates BST ordering:
#      5
#     / \
#    7   3
#
root = Node(5, Node(7), Node(3))
print(min_changes_to_bst(root))  # Output: 2

File: 46.js
class Node {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Node {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function minChangesToBST(root) {
  const inorderVals = [];

  function inorder(node) {
    if (!node) return;
    inorder(node.left);
    inorderVals.push(node.val);
    inorder(node.right);
  }

  inorder(root);

  const lis = [];

  function lowerBound(arr, x) {
    let l = 0, r = arr.length;
    while (l < r) {
      const mid = Math.floor((l + r) / 2);
      if (arr[mid] < x) l = mid + 1;
      else r = mid;
    }
    return l;
  }

  for (const val of inorderVals) {
    const idx = lowerBound(lis, val);
    if (idx === lis.length) lis.push(val);
    else lis[idx] = val;
  }

  return inorderVals.length - lis.length;
}


// Example usage:
const root = new Node(5, new Node(7), new Node(3));
console.log(minChangesToBST(root)); // 2
